server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -522,16 +530,46 @@ public Indexer(String indexName,
      */
     public void updateTargets(List<Reference> addedColumns) {
         // Store by ident for fast replacement.
-        Map<ColumnIdent, Reference> addedColumnsByIdent = new HashMap<>();
         for (Reference ref: addedColumns) {
-            addedColumnsByIdent.put(ref.column(), ref);
+            newColumns.put(ref.column(), ref);
         }
 
+        Function<ColumnIdent, Reference> newGetRef = columnIdent -> {
+            var ref = newColumns.get(columnIdent);
+            if (ref == null) {
+                ref = getRef.apply(columnIdent);
+            }
+            return ref;
+        };
+
         for (int i = 0; i < columns.size(); i++) {
-            Reference referenceWithOid = addedColumnsByIdent.get(columns.get(i).column());
-            if (referenceWithOid != null) {
-                columns.set(i, referenceWithOid);
+            var oldRef = columns.get(i);
+            if (oldRef.oid() == COLUMN_OID_UNASSIGNED) {
+                // try to resolve the column again, new reference may have been added to the table already or
+                // to the list of new columns returned by the {@link AddColumnTask}
+                Reference referenceWithOid = newGetRef.apply(oldRef.column());
+                if (referenceWithOid != null && referenceWithOid.oid() != COLUMN_OID_UNASSIGNED) {
+                    columns.set(i, referenceWithOid);
+                    valueIndexers.set(i, referenceWithOid.valueType().valueIndexer(
+                            referenceWithOid.ident().tableIdent(),
+                            referenceWithOid,
+                            getFieldType,
+                            newGetRef
+                    ));
+                    continue;
+                }
```

Comment:
Do I get it right, that we basically getting fresh state by combining 

1. Already known columns, and they are also fresh because `TransportShardUpsertAction` gets a new table from Schemas. (`Indexer.getRef` is fresh)
2. Getting new columns from `AddColumnResponse`.

I wonder if can avoid propagating `updateTargets` (and dropping it from `ValueIndexer`) by 
ensuring that `getRef` supplier which we already pass to `ObjectIndexer` is always fresh by:
1. Change `getRef` to get values from some new Map<Columnident, Reference> `Indexer.allColumnsByIdent`
2. When Indexer is created, we already ensure that table is refreshed so we can fill the map by actual data
3. Make Indexer.updateTargets update `allColumnsByIdent` map. (this is also related to another comment to cover races by taking ALL targets in AddColumnResponse)

Reply:
We must replace any `ValueIndexer` when a new column is finally added and contains on the needed `Oid`. The `ValueIndexer` will use a possible  `Oid` for generating the source and defining Lucene's field mapper.

Reply:
Ah yep, initial valueIndexers created from the ref without any oid, I missed that. Thanks for explaining!